x-pack/plugins/lens/public/indexpattern_plugin/indexpattern.tsx

```diff
@@ -9,36 +9,94 @@ import { render } from 'react-dom';
 import { Datasource, DataType } from '..';
 import { DatasourceDimensionPanelProps, DatasourceDataPanelProps } from '../types';
 
-interface IndexPatternPrivateState {
-  query: object;
+type OperationType = 'value' | 'terms' | 'date_histogram';
+
+interface IndexPatternColumn {
+  // Public
+  operationId: string;
+  label: string;
+  dataType: DataType;
+  isBucketed: false;
+
+  // Private
+  operationType: OperationType;
+  arguments: any;
+}
+
+export interface IndexPattern {
+  id: string;
+  fields: Field[];
+  title: string;
+  timeFieldName?: string;
+}
+
+export interface Field {
+  name: string;
+  type: string;
+  aggregatable: boolean;
+  searchable: boolean;
+}
+
+export interface IndexPatternPrivateState {
+  currentIndexPattern: string;
+
+  columnOrder: string[];
+  columns: {
+    [columnId: string]: IndexPatternColumn;
+  };
+
+  // Not persisted
+  indexPatterns: { [id: string]: IndexPattern };
 }
 
 // Not stateful. State is persisted to the frame
 export const indexPatternDatasource: Datasource<IndexPatternPrivateState> = {
-  async initialize() {
-    return { query: {} };
+  async initialize(state?: IndexPatternPrivateState | any) {
```

Comment:
`sth | any` just comes down to `any`, which allows to call this method with anything. The implementation doesn't check the value of state during runtime, so we are basically dropping all type safety guarantees here (e.g. how state is used inside of this function). We also discussed this in the chat, but IMO we should add a second generic type for this. I prepared this in a PR on this branch here: https://github.com/wylieconlon/kibana/pull/2/files

Gist:
The `getPersistableState` function returns a generic type `P`, and the `initialize` function will always be called with this generic type (or nothing if there isn't a saved object).

Reply:
Will update